draft-ietf-tls-esni.md

```diff
@@ -239,12 +239,12 @@ The `ECHConfigContents` structure contains the following fields:
 key_config
 : A `HpkeKeyConfig` structure carrying the configuration information associated
 with the HPKE public key. Note that this structure contains the `config_id`
-field, which applies to the entire ECHConfigContents. Sites MUST NOT publish
+field, which applies to the entire ECHConfigContents. Sites SHOULD NOT publish
 two different `ECHConfigContents` values with the same `HpkeKeyConfig` value.
 The RECOMMENDED technique for choosing `config_id` is to choose a random byte.
-This process is repeated if this config_id matches that of any valid ECHConfig,
-which could include any ECHConfig that has been recently removed from active
-use.
+This could lead to `config_id` collisions, especially if a backend server
+is provided with ECHConfig values by more than one client-facing server and
+if each of those support multiple KEMs, hence the SHOULD above.
```

Comment:
I don't think this example is right. If a backend server is supported by multiple client-facing servers (A and B), they need to set up their DNS properly, otherwise the connection will fail. For instance, multiple HTTPS RRs, one A's IP + A's ECHConfigList, and another with B's IP + B's  ECHConfigList.

Indeed this is why we switched from the old ECH DNS record: we got feedback from CDN operators that IPs and ECHConfigLists need to be correlated to avoid multi-CDN problems.

Reply:
"They need to setup their DNS properly" is the issue - it's the backend's DNS not the client-facing server's DNS.

Reply:
Yes, the backend and client-facing servers do need to work out how they delegate this information. This has nothing to do with config_ids. The keys are a property of the client-facing server, one that will rotate over time, and the backend's DNS needs to somehow present those keys.

If the backend server and client-facing server share enough infrastructure (Shared Mode), maybe the client-facing server's provisioning process is able to directly configures the backend server's DNS. More likely, I expect the backend server will delegate the name with a CNAME or AliasForm HTTPS record.

This is the same sort of deployment decisions folks already go through with A/AAAA records.

Reply:
Sorry, but doesn't the protocol need to support the case where the split-mode backend's DNS operator is independent of all client-facing servers? If not, I've a problem with that. If so, then we can get config_id collisions, even if those aren't much of a problem. (We are after all only arguing here about the example to illustrate a SHOULD so it's not the end of the world:-)

Reply:
> Sorry, but doesn't the protocol need to support the case where the split-mode backend's DNS operator is independent of all client-facing servers?

That's what CNAMEs and AliasForm records are for, no? Or do you mean independent as in without any pointers to the client-facing server at all? Even without ECHConfigs, you need to point at the client-facing server's A/AAAA records, no?

>  If so, then we can get config_id collisions, even if those aren't much of a problem?

Again, collisions across client-facing servers aren't relevant to the protocol at all. See my comment above. I think you are misunderstanding what this is even doing.

Reply:
Collisions across client-facing servers would be relevant if the MUST not publish remained. I don't believe I'm misunderstanding config_id thanks:-)

CNAMEs etc are one way of doing things and likely will be most used.  But a backend is not in a position to say to a likely much larger CDN - hey please change your config_id because that other big CDN used the same one;-) So if we want the protocol to allow the backend to choose any DNS operator (unrelated to all CDNs) then the MUST NOT was wrong but so was the assumption that one CDN could entirely avoid config_id collisions *in the DNS* via iteration at key gen time.

Reply:
Ah yeah, the sentence is just wrong all around. :-) It both should be a SHOULD, not a MUST, but the SHOULD is also wrong because it should be about an individual client-facing server, and not the backend at all.

(I'm reworking that text now, which hopefully avoids this. Will upload a PR when that's done.)